python/revng/cli/_commands/fetch_debuginfo/elf.py

```diff
@@ -0,0 +1,195 @@
+#!/usr/bin/env python3
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+import os
+import shutil
+import subprocess
+from pathlib import Path
+
+from elftools.elf.elffile import ELFFile
+from elftools.elf.sections import NoteSection
+
+from .common import download_file, log
+
+
+def parse_build_id(file):
+    for section in file.iter_sections():
+        if not isinstance(section, NoteSection):
+            continue
+        for note in section.iter_notes():
+            desc = note["n_desc"]
+            if note["n_type"] != "NT_GNU_BUILD_ID":
+                continue
+            return desc
+    return None
+
+
+def parse_debug_link(data):
+    dataptr = 0
+    while dataptr < len(data):
+        while dataptr < len(data) and not (32 <= data[dataptr] <= 127):
+            dataptr += 1
+
+        if dataptr >= len(data):
+            break
+
+        endptr = dataptr
+        while endptr < len(data) and data[endptr] != 0:
+            endptr += 1
+
+        # TODO: If we want to parse checksum as well, we should continue
+        # and set up `dataptr = endptr`.
+        return data[dataptr:endptr].decode("latin-1")
+
+    return None
+
+
+def check_dbgsym_packages(verbose):
+    global_debug_package_list_path = "/etc/apt/sources.list.d/ddebs.list"
+    if os.path.exists(global_debug_package_list_path):
+        log("dbgsym packages are enabled...", verbose)
+        return True
+
+    log(
+        "warning: Please enable dbgsym packages by following "
+        "https://wiki.ubuntu.com/Debug%20Symbol%20Packages",
+        verbose,
+    )
+    return False
+
+
+# Look for the debug info in the following order:
+#   1. in file itself
+#   2. .gnu_debuglink
+#   3. canonical places
+#   4. use Debuginfod like apporach by finding it on web
+#   5. apt debug packages
+def fetch_dwarf(file: ELFFile, file_path, urls, verbose):
+    log("Looking for Debugging Information for an ELF...", verbose)
+
+    # 1) Check file itself.
+    section = file.get_section_by_name(".debug_info")
+    if section is not None:
+        log("Binary has debug info within itself", verbose)
+        return file_path
```

Comment:
Can a file have both `.debug_info` and external debug information?

Reply:
I think no... I'd consider that as a hack and non-standard way. In theory, you can use ```dd``` and copy whatever part of the ELF you want, but I guess we don't want to count on that.

There are 2 ways of generating separate debug files (on Linux):

- Separate Debug FIles -- ```.debug``` files, most common way, but not Standardized (let's say supported as ```GNU``` extensions)
- Split DWARF and .debug_sup``` -- Standardized way of keeping separate debug info files; part of DWARF 5;
  - In the case of Split DWARF, we do have some DWARF sections, but they are marked with ```.dwo``` suffix (e.g. ```debug_info.dwo```)

Reply:
Also, I am not able to add support for Mac OS, since I don't have a machine, but they use ```dsym``` files, very similar approach to these ```.debug``` files.

Reply:
Let's postpone macOS.
OK, for the rest.